drivers/flash/flash_stm32_qspi.c

```diff
@@ -608,6 +682,133 @@ static int setup_pages_layout(const struct device *dev)
 }
 #endif /* CONFIG_FLASH_PAGE_LAYOUT */
 
+static int qspi_program_addr_4b(const struct device *dev)
+{
+	uint8_t reg;
+	int ret;
+
+	/* Program the flash memory to use 4 bytes addressing */
+	QSPI_CommandTypeDef cmd = {
+		.Instruction = SPI_NOR_CMD_4BA,
+		.InstructionMode = QSPI_INSTRUCTION_1_LINE,
+	};
+
+	ret = qspi_send_cmd(dev, &cmd);
+	if (ret) {
+		return ret;
+	}
+
+	/*
+	 * Read control register to verify if 4byte addressing mode
+	 * is enabled.
+	 */
+	cmd.Instruction = SPI_NOR_CMD_RDCR;
+	cmd.InstructionMode = QSPI_INSTRUCTION_1_LINE;
+	cmd.DataMode = QSPI_DATA_1_LINE;
+
+	ret = qspi_read_access(dev, &cmd, &reg, sizeof(reg));
+	if (!ret && !(reg & SPI_NOR_4BYTE_BIT)) {
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int qspi_read_status_register(const struct device *dev, uint8_t *reg)
+{
+	QSPI_CommandTypeDef cmd = {
+		.Instruction = SPI_NOR_CMD_RDSR,
+		.InstructionMode = QSPI_INSTRUCTION_1_LINE,
+		.DataMode = QSPI_DATA_1_LINE,
+	};
+
+	return qspi_read_access(dev, &cmd, reg, sizeof(*reg));
+}
+
+static int qspi_write_status_register(const struct device *dev, uint8_t reg)
+{
+	QSPI_CommandTypeDef cmd = {
+		.Instruction = SPI_NOR_CMD_WRSR,
+		.InstructionMode = QSPI_INSTRUCTION_1_LINE,
+		.DataMode = QSPI_DATA_1_LINE,
+	};
+
+	return qspi_write_access(dev, &cmd, &reg, sizeof(reg));
+}
+
+static int qspi_write_enable(const struct device *dev)
+{
+	uint8_t reg;
+	int ret;
+
+	QSPI_CommandTypeDef cmd_write_en = {
+		.Instruction = SPI_NOR_CMD_WREN,
+		.InstructionMode = QSPI_INSTRUCTION_1_LINE,
+	};
+
+	ret = qspi_send_cmd(dev, &cmd_write_en);
+	if (ret) {
+		return ret;
+	}
+
+	do {
+		ret = qspi_read_status_register(dev, &reg);
+	} while (!ret && !(reg & SPI_NOR_WEL_BIT));
+
+	return ret;
+}
+
+static int qspi_program_quad_io(const struct device *dev)
+{
+	struct flash_stm32_qspi_data *data = DEV_DATA(dev);
+	uint8_t reg;
+	int ret;
+
+	/* Check if QE bit setting is required */
+	ret = qspi_read_status_register(dev, &reg);
+	if (ret) {
+		return ret;
+	}
+
+	/* Quit early when QE bit is already set */
+	if (reg & SPI_NOR_QE_BIT) {
+		goto out;
+	}
+
+	ret = qspi_write_enable(dev);
+	if (ret) {
+		return ret;
+	}
+	reg |= SPI_NOR_QE_BIT;
+
+	ret = qspi_write_status_register(dev, reg);
```

Comment:
Just cosmetic: you may want remove empty line at 783 and add one above 782; the bit set is rather prerequisite for the `qspi_write_status_register` than post operation of the `qspi_write_enable`. It is a little bit dangling there.